### ISSUE TITLE
Proposal: expose pid of the container's process with pid-file option

### ISSUE THREAD

**Issue by dqminh** [docker/docker#25380]
For simple tools that want to start a docker container with a service manager (systemd ahem.), it's useful to have it monitor  a pid file (like `runc run --pid-file`) especially now that we have `--live-restore`

The proposal is to add `--pid-file` support to `docker run` and `docker start` that will probably just proxied all the way to runc.

**Comment by mlaventure** [comment]
I'm not sure I understand.

If a service manager wants to start a container without docker, it would probably use directly a runntime like `runc` and can directly pass the `--pid-file` flag to it.

If it starts the containers via docker then it can get the PID via `inspect` atm. But, IMHO, if docker starts the containers, docker should be the one taking care of them not the service manager.

/cc @tonistiigi

**Comment by tonistiigi** [comment]
@mlaventure I guess it is just about docker daemon writing same pid number that is exposed by inspect to a file as well because this is a common pattern. I'd be ok with it if you can show a use case where `docker inspect` can't be run(and written to file if needed) before using the pid-file.

**Comment by dqminh** [comment]
@mlaventure its possible to use runc directly but then you lose other benefits of docker such as image management, network and volumes setup etc.

@tonistiigi it's always possible to do `docker inspect` then write out the pid file, it's just not convenient to do so.

**Comment by dqminh** [comment]
@tonistiigi @mlaventure gentle nudge for any thoughts on whether this is worthwhile ?

**Comment by crosbymichael** [comment]
my 2c.  at the docker level the abstraction is above individual pids.  Its not that important for a user to focus on.

**Comment by mlaventure** [comment]
@dqminh Sorry, I'm still not convinced this is needed as `inspect` should do the job and I rather not add an extra option in that case.

**Comment by crosbymichael** [comment]
Lets not expose this right now.  If we have a few other people providing usecases for this type of feature we can look at it again.  Pids is one thing we want to kinda abstract away and have the container be the unit that people interact with in the engine.

**Comment by Mandrake981** [comment]
I am also looking into this. Using applications to monitor Docker containers (such as Monit) requires there to be a PID file for whatever we're wanting to monitor. If the PID file could be written (when the container is started), it would save all of the bogus hokus pokus drudgery needed to complete this. Monit would essentially monitor the PID in the PID file (while also checking the connection with every update), and it could be used to stop and restart a container should something happen.

In our case, our application (an environmental program written for the 3 branches of our military) is being developed, and we've ran into a snag where Oracle 12c keeps throwing up its hands and saying "I quit" (too many open cursors), so essentially the container dies (our application serves up a 500 error when it cannot get to the Oracle database), but that's not considered "dead" by Docker (it's doing what it can do under the circumstances). A restart would fix the application (cause the connection to be reset between the Oracle database and the Docker container), but it sees nothing wrong, so cannot fix it. That being said, if the PID file were there, I could set up Monit to monitor the container (checking the connection every few minutes), and if it fails (gives a 500 error, etc.), then I could have it restart the container (which does fix the issue). Since we have not ironed out the problems that have come up with our switch to 12c, we're using OJDBC settings just a little above what our developers (who handled the conversion - where we went from Weblogic to Glassfish) are using. Finding "production" ready info for these things are about impossible, and at least if we had the application up we could test it and make adjustments without it being a problem.
